### ISSUE TITLE
[5.1] Can't flush tags on Redis cluster cache after updating to Laravel 5.1.45

### ISSUE THREAD

**Issue by mattmant** [laravel/framework#17713]
- Laravel Version: 5.1.45
- PHP Version: 5.5.9
- Redis Version: 3.2.6

### Description:
I updated Laravel from version 5.1.20 to version 5.1.45. I'm using cache Redis with tags to store some informations. The application is running on 3 web-servers and Redis is in cluster mode.

After the update I can't flush tags saved in Redis because I get the following error: 
```
NotSupportedException in RedisCluster.php line 380:
Cannot use 'DEL' with redis-cluster.
```

The important thing is that before updating Laravel the application was stable and was working well.

I saw that there was many important changes into `Illuminate\Cache\RedisTaggedCache` with the introduction of `$this->deleteStandardKeys();` into the `flush()` method.

### Steps To Reproduce:

You can test the behavior in your local machine with a single Redis node (the same happens if you have a real Redis cluster also).

1) Set Redis in cluster mode by putting `'cluster' => true` in your Redis config in `config/database.php `. For example:
```
'redis' => [
    'cluster' => true,
    'default' => [
        'host' => '127.0.0.1',
        'port' => 6379,
        'database' => 0
    ]
]
```

2) Store at least two keys for the same tag:
```
Cache::store( "redis" )->tags( "tag" )->put( "key1", "test", 100 );
Cache::store( "redis" )->tags( "tag" )->put( "key2", "test", 100 );
```

3) Now try to flush the tag:
`Cache::store( "redis" )->tags( "tag" )->flush();`

You will get the following red message:
`Predis\NotSupportedException with message 'Cannot use 'DEL' with redis-cluster.'`

**Comment by mattmant** [comment]
Update: the issue was introduced with the `5.1.30` release of `laravel/framework`.

**Comment by taylorotwell** [comment]
You are seeing a Predis exception. I'm not sure there is much we can do about it? Perhaps asks on their repository?

**Comment by Eldair** [comment]
Same thing with `'cluster' => false` in 5.4

**Comment by mattmant** [comment]
Yes, it is a Predis exception and it would be a good idea to ask in their repository, after have reproduced the behaviour directly on it.

But, updating Laravel from 5.1.20 to 5.1.45 didn't affect Predis package version: it still remains v1.1.1.
It seems that after changes on `Illuminate\Cache\RedisTaggedCache` it doesn't properly support redis cluster... because my application was stable and was working well before these changes.

**Comment by calebfavor** [comment]
Confirmed broken on 5.4 even with cluster false like @Eldair mentioned.

```
Cache::store( "redis" )->tags( "mytag" )->put( "key", "value", 100 );
Cache::store( "redis" )->tags( "mytag" )->flush();

Cache::store( "redis" )->tags( "mytag" )->get( "key" ); // returns true
```

Tested with:
laravel: v5.4.10
predis: v1.1.1
redis server: v3.2.7

Will do some digging later today to see if its an easy fix.

**Comment by mattmant** [comment]
This issue should be tagged for any version of Laravel because of wrong definition of tags into `Illuminate\Cache\RedisTaggedCache`.

Refer to this https://github.com/laravel/framework/pull/17792#issuecomment-277977633 for more informations, and this about what could be changed: https://github.com/laravel/framework/pull/17792/files.

**Comment by tillkruss** [comment]
Your updated #17792 PR that uses hash tags looks good to me. If it doesn't get merged into 5.1, I'd at least submit it for 5.4.

**Comment by Eldair** [comment]
Is this moving forward?

**Comment by JeramyNS** [comment]
Still no fix or workaround for this bug?

**Comment by benjamincrozat** [comment]
So it seems like this bug is the same as the one I posted (#22788). Does anybody could take a look? @taylorotwell @themsaid

**Comment by tillkruss** [comment]
I'd suggest to submit a PR for v5.6.

**Comment by benjamincrozat** [comment]
Will do that once I have some time. I'm not really familiar with the code base.

**Comment by woufapp** [comment]
Had the issue too. Have the same app on a redis 3.2.8 don't reproduce the error.

**Comment by rikless** [comment]
In my case, it was a bad definition in config/database.php.

I think most of us are using an upgraded version of Laravel. And it seems that config file definition has changed.

Open your config/database.php file, you may have this : 

```php
'redis' => [

        'client' => 'predis',

        'options' => [
            'cluster' => 'redis',
        ],

        'clusters' => [
            'default' => [
                [
                    'host' => env('REDIS_HOST', '127.0.0.1'),
                    'password' => env('REDIS_PASSWORD', null),
                    'port' => env('REDIS_PORT', 6379),
                    'database' => 0,
                ],
            ],
        ],

    ],
```

Replace it by : 

```php
'redis' => [

        'client' => 'predis',
        'default' =>
            [
                'host' => env('REDIS_HOST', '127.0.0.1'),
                'password' => env('REDIS_PASSWORD', null),
                'port' => env('REDIS_PORT', 6379),
                'database' => 0,
            ],

    ],
```

**Comment by T0miii** [comment]
Any update on this?

Getting similar error wthile trying to flush a tag. Also using Redis Clusters.

**Comment by madsem** [comment]
@mattmant I wonder how this still isn't merged :(
Is there any alternative right now to be able to do Cache::tags('something')->flush() with Redis Cluster enabled?

**Comment by GrahamCampbell** [comment]
> I wonder how this still isn't merged :(

This is not a pull request. There's nothing to merge?

**Comment by madsem** [comment]
Okay sorry for not using 100% accurate verbiage @GrahamCampbell , but in the end this still doesn't work in 5.6 and cannot flush tagged cache with redis cluster.

**Comment by GrahamCampbell** [comment]
We're open to pull requests to fix this.

**Comment by madsem** [comment]
That's what I meant, there was one a year ago for 5.1 https://github.com/laravel/framework/pull/17792/files 

;)

**Comment by devcircus** [comment]
Taylor mentioned on that pull request, that something similar may be considered for master. Maybe someone just needs to PR it?

**Comment by tillkruss** [comment]
Yeah, just send a PR to v5.7.

**Comment by danygiguere** [comment]
I haven't found a solution but maybe a way around it for some cases. In my case I was tagging single posts with this tag: `Cache::tags(["post_$postId"])`and I had a problem when I was trying to clear the cache for a user’s posts like so 

> foreach($user->posts() as $post) {
>                 Cache::tags(["post_$post->id"])->flush();
>             } 

Somehow the bug didn't occurred all the time. Within some methods the foreach worked properly. Only in one method it failed.

I found a solution by adding this tag `"post_user_$post->user->id"`
So now I have two tags:

> Cache::tags(["post_user_$post->user->id", "post_excerpt_$postId"])

 and I can clear the cache for those posts with: 
`Cache::tags(["post_user_$user->id"])->flush();`

**Comment by cameronpalatas** [comment]
Any update on this? Running into the same issue.

**Comment by laurencei** [comment]
Can anyone confirm this is occuring on Laravel 5.7 (or 5.5)? Would be good to know if any of the recent updates fixed this?

**Comment by zeetabit** [comment]
also that error :-(

**Comment by cg-rsands** [comment]
Same problem, anyone any idea?? @taylorotwell

**Comment by xVismutx** [comment]
So, how to use tags with this bug? any way around it?

**Comment by IbrahimFathy19** [comment]
Any updates on this issue?
if not, then I'm ready to work on it.

**Comment by Jafo232** [comment]
This is still an issue with Memcached..

**Comment by driesvints** [comment]
@Jafo232 we're welcoming prs to help solve this.

**Comment by taylorotwell** [comment]
It's still an issue because nobody has fixed it. Simple as that. 😄 **Anyone** can fix it by making a PR to the framework.

**Comment by Nuranto** [comment]
I have the issue with memcached too.

**Comment by mattmant** [comment]
> I have confirmed this is working on a Redis cluster I provisioned on ScaleGrid.io and using the `phpredis` PHP extension (default Redis driver for Laravel now). If it is not working on Predis I'm not sure what to do about that. Predis is no longer maintained as a library.

As I have suggested three years ago, I'm still using the following solution, without any side effects on performance/data: https://github.com/laravel/framework/pull/17792/files

Refer to this https://github.com/laravel/framework/pull/17792#issuecomment-277977633 for more information.
